stm32n6: Fix various cases around flashing

erwango opened the pull request:
Fixes https://github.com/zephyrproject-rtos/zephyr/issues/85041 for non STM32N6 platforms
Fixes https://github.com/zephyrproject-rtos/zephyr/issues/85207 to better inform users when they miss the appropriate configuration

erwango committed:
scripts: runners: stm32cubeprogrammer: Review f/w selection algorithm

Aim is to avoid 2 issues:
- Requesting --download-address to flash hex files
- Flashing hex files (non signed) on N6 which works but doesn't allow
persistent firmware

Hence this changes binds --download-address with usage of .bin files.
If --download-address was not provided, default to hex flashing.

File existence done with isfile() will ensure the required file
is available before flashing and report an error if this is not the
case.
Note: In specific N6 case, we're verifying zephyr.signed.bin is
available instead of zephyr.bin. This is ensure since self.cfg.bin_file
matches runners_yaml_props_target configuration (set to
zephyr.signed.bin on STM32N6).

Signed-off-by: Erwan Gouriou <erwan.gouriou@st.com>

gautierg-st commented on soc/st/stm32/stm32n6x/CMakeLists.txt:
```diff
@@ -21,10 +21,7 @@ if(SIGNING_TOOL_FIND STREQUAL SIGNING_TOOL_FIND-NOTFOUND)
   Signing Image tool (${SIGNING_TOOL}) is not available.
   Signed image will not be generated.
   You won't be able to run image on board.")
```
Maybe add an explicit message: "Add STM32_SigningTool_CLI to your $PATH" ?

gautierg-st reviewed (commented):
I wonder if we could provide a default path to the signing tool in case it is not found.
On windows, this would be "C:\Program Files\STMicroelectronics\STM32Cube\STM32CubeProgrammer\bin"

erwango committed:
scripts: test runners: stm32cubeprogrammer: Update --download-address test

Fix --download-address to expect a signed.bin file when --download-address
is provided.

To allow testing this case despite the checks put in the runner, add a
patch functions which return that zephyr.bin is actually present

Signed-off-by: Erwan Gouriou <erwan.gouriou@st.com>

erwango commented:
@ajarmouni-st FYI

erwango commented on scripts/west_commands/runners/stm32cubeprogrammer.py:
```diff
@@ -169,7 +169,7 @@ def do_add_parser(cls, parser):
             type=functools.wraps(int)(lambda s: int(s, base=0)),
             required=False,
-            help="Address where flashing should be done"
+            help="Address where flashing should be done. To be used only for .bin files"
```
Fixed

mathieuchopstm commented on scripts/west_commands/runners/stm32cubeprogrammer.py:
```diff
@@ -271,15 +275,20 @@ def flash(self, **kwargs) -> None:
-        elif self.cfg.bin_file is not None and os.path.isfile(self.cfg.bin_file) and \
-            "zephyr.signed" in self.cfg.bin_file:
+        elif self.cfg.bin_file is not None and self._download_address is not None:
+            # Use bin file if available and --download-address provided
```
```suggestion
            # Use bin file if a signed binary is available
            # --download-address must be provided
```
If `.bin` is provided but not `--download-address`, maybe a warning should be logged too?

mathieuchopstm commented on scripts/west_commands/runners/stm32cubeprogrammer.py:
```diff
@@ -169,7 +169,7 @@ def do_add_parser(cls, parser):
             type=functools.wraps(int)(lambda s: int(s, base=0)),
             required=False,
-            help="Address where flashing should be done"
+            help="Flashing location address. To be used only with .bin files"
```
```suggestion
            help="Flashing location address. Required to flash .bin files, ignored otherwise."
```

I think the following wording could be better:
```suggestion
            help="Address where binary should be flashed. Required to flash .bin files, ignored otherwise."
``````

mathieuchopstm reviewed (commented):
Minor nits about wording, LGTM functionally.

erwango committed:
scripts: runners: stm32cubeprogrammer: Factorize type lambda function

Now that it is used twice, factorize lambda and comment.

Signed-off-by: Erwan Gouriou <erwan.gouriou@st.com>

erwango commented:
New iteration with reviewed binary selection logic:
Use .bin if download-address provided.
.bin file presence check is done with with `os.path.isfile()` which verify `self.cfg.bin_file` presence.
`self.cfg.bin_file` being configured using runners_yaml_props_target as zephyr.signed.bin on STM32N6.

Thanks @mathieuchopstm for the hint.

mathieuchopstm commented on scripts/west_commands/runners/stm32cubeprogrammer.py:
```diff
@@ -271,15 +269,18 @@ def flash(self, **kwargs) -> None:
-        elif self.cfg.hex_file is not None and os.path.isfile(self.cfg.hex_file):
-            # --user-elf not used and no bin file given, default to hex
+        elif self.cfg.hex_file is not None:
+            # --use-elf not used and no signed bin file given, default to hex
```
That's what I was thinking about:
```suggestion
            # Neither --use-elf nor --download-address are present:
            # default to flashing using hex file.
```

mathieuchopstm commented on scripts/west_commands/runners/stm32cubeprogrammer.py:
```diff
@@ -146,6 +146,10 @@ def capabilities(cls):
     def do_add_parser(cls, parser):
+        # To accept arguments in hex format, a wrapper lambda around int() must be used.
+        # Wrapping the lambda with functools.wraps() makes it so that 'invalid int value'
+        # is displayed when an invalid value is provided for this argument.
```
Nit:
```suggestion
        # is displayed when an invalid value is provided for these arguments.
```

erwango commented on scripts/west_commands/tests/test_stm32cubeprogrammer.py:
```diff
@@ -447,6 +449,12 @@
 )
 """Test cases."""
 
+os_path_isfile = os.path.isfile
+
+def os_path_isfile_patch(filename):
+    if filename == RC_KERNEL_BIN:
+        return True
+    return os_path_isfile(filename)
```
Yes, not sure why but a stub is also required for `RC_KERNEL_BIN`.

JarmouniA reviewed (approved):
Thanks @erwango

erwango commented:
@pdgendt Mind having a look ?
